### ISSUE TITLE
netconfig.managed state.apply unexpected behaviour of test & debug variables

### ISSUE THREAD

**Issue by malbertus** [saltstack/salt#44847]
### Description of Issue/Question

Running a state.apply or state.sls with netconfig.managed with test and debug variables results in unexpected behaviour. Test: True variable set in SLS seems to be ignored, while debug=true variable set on CLI seems to be ignored. 

When bypassing state module and calling net.load_config directly using cli, behaviour is as expected.

### Setup

Vagrant environment with Salt Master/Minion/Proxy and vSRX. Custom ext_pillar set up to IPAM using http.query.

```
init.sls
include:
    - config_juniper.configure
```

```
configure.sls
configure:
    netconfig.managed:
      - template_name: salt://templates/junos.jinja
      - network: {{salt['pillar.get']('network')}}
      - device: {{salt['pillar.get']('device')}}
      - ip: {{salt['pillar.get']('ip')}}
```

### Steps to Reproduce Issue

Baseline, neither test nor debug applied and state applies as expected.
```
ubuntu@saltmaster:~$ sudo salt 'srx1' state.apply
srx1:
----------
          ID: configure
    Function: netconfig.managed
      Result: True
     Comment: Configuration changed!
     Started: 04:58:54.090764
    Duration: 5571.34 ms
     Changes:   
              ----------
              diff:
                  [edit system]
                  -  host-name baseline.lab;
                  +  host-name srx1.lab;

Summary for srx1
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   5.571 s
```
Adding the following variables to configure.sls:
test: True
debug: True

Appears as though debug is working (loaded_config is displayed) but test mode is not working (configuration change occurred).

```
ubuntu@saltmaster:~$ sudo salt 'srx1' state.apply
srx1:
----------
          ID: configure
    Function: netconfig.managed
      Result: True
     Comment: Configuration changed!
     Started: 05:03:12.328063
    Duration: 5504.173 ms
     Changes:   
              ----------
              diff:
                  [edit system]
                  -  host-name baseline.lab;
                  +  host-name srx1.lab;
              loaded_config:
                  system {
                      host-name srx1.lab;
                  }

Summary for srx1
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   5.504 s
```

Remove 'test: True' from configure.sls but add to the command line, however 'debug: True' left in SLS. Works as expected.

```
ubuntu@saltmaster:~$ sudo salt 'srx1' state.apply test=True
srx1:
----------
          ID: configure
    Function: netconfig.managed
      Result: None
     Comment: Configuration discarded.
              
              Configuration diff:
              
              [edit system]
              -  host-name baseline.lab;
              +  host-name srx1.lab;
              
              Loaded config:
              
              system {
                  host-name srx1.lab;
              }
              
     Started: 05:06:57.948617
    Duration: 1082.001 ms
     Changes:   

Summary for srx1
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   1.082 s

```

Remove 'debug: True' from configure.sls, and apply instead on the cli along with test=True. Loaded_config disappears, doesn't appear that debug mode is running. Test mode working

```
ubuntu@saltmaster:~$ sudo salt 'srx1' state.apply test=True debug=True
srx1:
----------
          ID: configure
    Function: netconfig.managed
      Result: None
     Comment: Configuration discarded.
              
              Configuration diff:
              
              [edit system]
              -  host-name baseline.lab;
              +  host-name srx1.lab;
     Started: 05:10:20.753379
    Duration: 1149.269 ms
     Changes:   

Summary for srx1
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   1.149 s
```

Final test case - test: True applied to SLS and debug=True applied via CLI. Neither works

```
ubuntu@saltmaster:~$ sudo salt 'srx1' state.apply debug=True
srx1:
----------
          ID: configure
    Function: netconfig.managed
      Result: True
     Comment: Configuration changed!
     Started: 05:16:18.180135
    Duration: 5281.746 ms
     Changes:   
              ----------
              diff:
                  [edit system]
                  -  host-name baseline.lab;
                  +  host-name srx1.lab;

Summary for srx1
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   5.282 s
```

### Versions Report

ubuntu@saltmaster:~$ salt --versions-report
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: 1.11.2
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 20 2017, 18:23:56)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-101-generic
         system: Linux
        version: Ubuntu 16.04 xenial

**Comment by garethgreenaway** [comment]
@malbertus Thanks for the report.  It appears that the `test` argument is not currently supported by the `netconfig` state module.  Tagging @mirceaulinic who has written a good portion of that particular state module.

**Comment by mirceaulinic** [comment]
Hi @malbertus - thanks for opening this.

It is something I am aware of, but you did good reporting it. This shouldn't be a blocker, but I will try to get this fixed ASAP.

Some key points to remember (mostly notes to self):

- the `state.sls`, `state.apply` functions and so on, don't have a `debug` argument, see for example https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.state.html#salt.modules.state.sls. But we can work this out from `kwargs`. Meanwhile, if you need this (and it's a nice feature to use when the state is under development), you can leave `debug: true` in your formula.

- The `test` one is overriden by default by the Salt State system, and might be very difficult or impossible to pass it from the SLS formula, and we might need to rename it. But I will investigate this closer and check if I can find a way so you can continue having `test: true`. Meanwhile, you can either pass `test=True` from the CLI, or add `test: true` in the config file (typically `/etc/salt/proxy`).

**Comment by mirceaulinic** [comment]
Just an update here: I tried twice to work this out, but it looks like a dead end, and we should probably just remove the `debug=True` CLI argument from the examples. I'm not sure if we can make this work fine.
I had the feeling that in past releases you could fine the CLI arguments in a `fun_args` or something similar under `__opts__`, but apparently I was wrong.
Meanwhile, the explanation above should suffice for any use-case.
I'm not giving up on this yet, but I can't currently think of a workaround.

**Comment by mirceaulinic** [comment]
I think I have figured out a different way to tackle this, though it would require a small syntax change: https://github.com/saltstack/salt/pull/47246.

**Comment by rallytime** [comment]
Closed via #47246
